Update Custom Checkout types

danwang-stripe opened the pull request:
### Summary & motivation

r? @pololi-stripe 

Update Custom Checkout types for custom_checkout_beta_2

* Added lineItem.amountsSubtotal, amountDiscount, amountTaxInclusive, amountTaxExclusive
* Removed lineItem.amount
* Added lineItem.recurring.usageType, isProrated, intervalCount
* Removed lineItem.recurring.interval_count
* Added total.appliedBalance and balanceAppliedToNextInvoice
* Added tax
* Added status
* Added recurring
* Added lastPaymentError
* Added discount.recurring
* Added lineItem.adjustableQuantity

### Testing & documentation

<!-- How did you test this change? This can be as simple as "I wrote unit tests...". -->

<!-- If this is an API change, have you updated the documentation? -->

<!-- OTHER: Consider checking "Allow edits from maintainers" below. -->

pololi-stripe commented on types/stripe-js/custom-checkout.d.ts:
```diff
@@ -188,13 +207,54 @@ export type StripeCustomCheckoutAddressElementOptions = {
   ): StripeAddressElement | null;
+  getElement(
+    elementType: 'expressCheckout'
+  ): StripeExpressCheckoutElement | null;
```
The CustomCheckout ECE has a slight different interface from this.

pololi-stripe commented on types/stripe-js/custom-checkout.d.ts:
```diff
@@ -203,18 +263,8 @@ export interface StripeCustomCheckoutElementsActions {
+  createElement(
+    elementType: 'expressCheckout',
+    options: StripeCustomCheckoutExpressCheckoutElementOptions
+  ): StripeExpressCheckoutElement;
```
Similar to the above comment, CustomCheckout ECE has a different interface.

danwang-stripe commented:
r? @pololi-stripe 

Updated in b475dd903dc708b92a58b29687ca69beb5e0bcc1 and f175d7e

* Added line item adjustable quantity (forgot this earlier)
* Updated type for ECE. To do this, I split all the ECE events into separate interfaces, then recombined them in the ECE type, and selected only the applicable events for Custom Checkout.

danwang-stripe commented:
r? @pololi-stripe 

I ditched the strategy of trying to reuse existing definitions. Instead, I just copied in the on/once/off methods for all events that we need, changing the return type in e80ef80c8738643fa5a9218b0df7a2f8695a4474

pololi-stripe commented on types/stripe-js/custom-checkout.d.ts:
```diff
@@ -188,13 +216,177 @@ export type StripeCustomCheckoutAddressElementOptions = {
+
+export type StripeCustomCheckoutExpressCheckoutElement = StripeElementBase & {
+  /**
+   * Triggered when the element is fully rendered and can accept `element.focus` calls.
```
```suggestion
   * Triggered when the element is fully rendered.
```

pololi-stripe reviewed (commented):
one minor comments
